toqito/matrix_ops/tensor_comb.py

```diff
@@ -30,42 +35,66 @@ def tensor_comb(states: list[np.ndarray], k: int) -> dict:
 
     .. jupyter-execute::
 
-        import numpy as np
-        from toqito.matrix_ops import tensor_comb
+     from toqito.matrix_ops import tensor_comb
+     import numpy as np
 
-        e_0 = np.array([1, 0])
-        e_1 = np.array([0, 1])
-        result = tensor_comb([e_0, e_1], 2)
+     e_0 = np.array([1, 0])
+     e_1 = np.array([0, 1])
 
-        for key, mat in result.items():
-            print(f"tensor_comb{key} =\n{mat}\n")
+     result = tensor_comb([e_0, e_1], 2, mode="injective", density_matrix=False)
+
+     for key, mat in result.items():
+         print(f"tensor_comb{key} =\n{mat}\n")
 
     References
     ==========
     .. bibliography::
         :filter: docname in docnames
 
     :raises ValueError: If the input list of states is empty.
-    :param states: A list of quantum state vectors represented as numpy arrays.
-    :param k: The length of the sequence for generating tensor products.
-    :return: A dictionary where:
-
-        - Keys represent sequences (as tuples) of quantum state indices,
-        - Values are density matrices corresponding to the tensor product of
-          the state vectors for the sequence.
+    :param states: A list of state vectors.
+    :param k: The length of the sequence.
+    :param mode: Determines the type of sequences. Default is :code:`"injective"`.
+           ``non-injective`` will allow repetitions in sequences,
+           ``injective`` will ensures sequences are injective (no repetitions) and
+           ``diagonal`` will allow sequences with repeated indices (diagonal elements).
+    :param density_matrix: Determines whether the return is a density matrix or a ket.
+            Default is ``True``.
+
+    :return: A dictionary where keys are tuples representing sequences of state indices,
+        and values are density matrices of the tensor products of the corresponding
+        state vectors or tensor products of the corresponding state vectors based on
+        input :code:`density_matrix` being either ``True`` or ``False``.
 
     """
     if not states:
         raise ValueError("Input list of states cannot be empty.")
 
-    possible_sequences = list(itertools.product(range(len(states)), repeat=k))
+    if mode not in ("injective", "non-injective", "diagonal"):
+        raise ValueError("mode must be injective, non-injective, or diagonal.")
+
+    if mode == "injective" and k > len(states):
+        raise ValueError("k must be less than or equal to the number of states for injective sequences.")
+
+    # Generate sequences based on the selected mode.
+    match mode:
+        case "injective":
+            sequences = list(itertools.permutations(range(len(states)), k))
+        case "non-injective":
+            sequences = list(itertools.product(range(len(states)), repeat=k))
+        case "diagonal":
+            sequences = [(i,) * k for i in range(len(states))]
```

Comment:
not sure why this line is showing up as uncovered by unit tests when there is a test for it. 

![image](https://github.com/user-attachments/assets/4420328e-fc24-45cc-b335-3bf84aa44a4e)

Reply:
Yeah, that's odd. I've noticed that codecov can be a bit finicky sometimes and I wonder if that's perhaps the case here.